Adds a feature of single member restoration in multi-node etcd.

ishan16696 opened the pull request:
**What this PR does / why we need it**:
This PR adds a feature for restoration of single member in mult-node etcd cluster in case of data/data-dir of etcd member found to be corrupt/invalid.

**Which issue(s) this PR fixes**:
Fixes  https://github.com/gardener/etcd-druid/issues/368

**Special notes for your reviewer**:
Design Doc PR: https://github.com/gardener/etcd-druid/pull/376

**Release note**:
```noteworthy user
For multi-node etcd: Added a feature of single member etcd restoration in case of data/data-dir of etcd member found to be corrupted or invalid.
```

ishan16696 commented:
Please note: etcd-events case(when no snapstore is configured) is still needs to be handled in multi-node etcd-events.
/do-not-merge

ishan16696 commented on pkg/initializer/initializer.go:
```diff
@@ -225,3 +230,33 @@ func (e *EtcdInitializer) removeDir(dirname string) error {
+// * Remove the member from the cluster
+// * Clean the data-dir of member that needs to be restored.
+// * Add a new member as a learner(non-voting member)
+// * Start a learner by providing config to etcd
+// * Promote the learner to voting member.
```
yes but I added it as for completion of steps . I will remove it. No strong opinion on this.

shreyas-s-rao commented on pkg/miscellaneous/miscellaneous.go:
```diff
@@ -462,5 +464,39 @@ func GetInitialClusterState(ctx context.Context, logger logrus.Entry, clientSet
+		logger.Infof("Member %v with [ID: %v] has been promoted", member.GetName(), strconv.FormatUint(member.GetID(), 16))
+		return nil
+	} else if errored.Is(err, rpctypes.Error(rpctypes.ErrGRPCMemberNotLearner)) {
+		//Member is not a learner
```
Is there a different way to ascertain that the member is a voting member of the cluster? If at all in the future, etcd community introduces a new kind of member which is neither `learner` nor `voting member`, then this check will not work.

ishan16696 replied:
we are currently using `etcd v3.4.13` and we don't see any plan to move  `v3.5.x` or `v3.6.x` anytime soon right ?
And even in etcd `v3.5.x` or `3.6.x` I don't see any new concept of new kindof member. so spending time on something which will not come anytime soon (for almost 2+ yrs) ... seems very unproductive to me right now.

shreyas-s-rao commented on pkg/miscellaneous/miscellaneous.go:
```diff
@@ -462,5 +464,39 @@ func GetInitialClusterState(ctx context.Context, logger logrus.Entry, clientSet
+}
+
+// CheckLearner checks whether a learner(non-voting) member present or not.
+func CheckLearner(ctx context.Context, cli etcdClient.ClusterCloser) (bool, error) {
```
Change name to `CheckForLearner` or `CheckIfLearnerPresent`. Maybe also return the learner's member ID, if required later on.

shreyas-s-rao commented on pkg/miscellaneous/miscellaneous_test.go:
```diff
@@ -343,9 +343,10 @@ var _ = Describe("Miscellaneous Tests", func() {
 		)
 		Context("In single-node etcd", func() {
-			It("Should return the cluster state as `new` ", func() {
+			It("Should return the cluster state as empty string ", func() {
```
Shouldn't the cluster state still be `new` for this case?

ishan16696 replied:
Ideally this test case should be removed as I added this when we are depending on [GetInitialClusterState](https://github.com/gardener/etcd-backup-restore/blob/master/pkg/miscellaneous/miscellaneous.go#L449) to get state but now it will only determine state for scale-up scenario and I have also  rename GetInitialClusterState to `GetInitialClusterStateIfScaleup` to avoid ambiguity.

shreyas-s-rao replied:
Ok. In that case, can you please adapt these tests too? I would like to avoid removing existing test cases if possible.

ishan16696 replied:
done.

shreyas-s-rao commented on pkg/miscellaneous/miscellaneous_test.go:
```diff
@@ -368,11 +369,11 @@ var _ = Describe("Miscellaneous Tests", func() {
 		})
 		Context("In multi-node etcd bootstrap", func() {
-			It("Should return the cluster state as `new` ", func() {
+			It("Should return the cluster state as empty string ", func() {
```
Same as above

ishan16696 replied:
no, this is correct

shreyas-s-rao commented on pkg/server/httpAPI.go:
```diff
@@ -466,6 +476,32 @@ func (h *HTTPHandler) serveConfig(rw http.ResponseWriter, req *http.Request) {
+
+		// check for a learner presence in a cluster.
+		// if a learner is present then return ClusterStateExisting else ClusterStateNew.
+		present, _ := m.IsLearnerPresent(ctx)
```
Can you please also read the error from the function call and act on it accordingly?

ishan16696 replied:
yes, will do

ishan16696 replied:
done. PTAL

shreyas-s-rao commented on pkg/server/httpAPI.go:
```diff
@@ -466,6 +476,32 @@ func (h *HTTPHandler) serveConfig(rw http.ResponseWriter, req *http.Request) {
 
+// GetClusterState returns the Cluster state either `new` or `existing`.
+func (h *HTTPHandler) GetClusterState(ctx context.Context, clusterSize int, clientSet client.Client, podName string, podNS string) string {
+	if clusterSize == 1 {
```
Shouldn't this case be handled within `GetInitialClusterState` method?

ishan16696 replied:
I have re-name `GetInitialClusterState` the func to avoid ambiguity.

shreyas-s-rao commented on pkg/server/httpAPI.go:
```diff
@@ -466,6 +476,32 @@ func (h *HTTPHandler) serveConfig(rw http.ResponseWriter, req *http.Request) {
+	// clusterSize > 1
+	state := miscellaneous.GetInitialClusterState(ctx, *h.Logger, clientSet, podName, podNS)
+	if len(state) == 0 {
+		// Not a Scalup scenario.
```
```suggestion
		// Not a Scale-up scenario.
```

shreyas-s-rao commented on pkg/server/httpAPI.go:
```diff
@@ -466,6 +476,32 @@ func (h *HTTPHandler) serveConfig(rw http.ResponseWriter, req *http.Request) {
+	state := miscellaneous.GetInitialClusterState(ctx, *h.Logger, clientSet, podName, podNS)
+	if len(state) == 0 {
+		// Not a Scalup scenario.
+		// Either a multi-node bootstarp or a restoration of single member in multi-node.
```
```suggestion
		// Either a multi-node bootstrap or a restoration of single member in multi-node.
```

shreyas-s-rao commented on pkg/server/httpAPI.go:
```diff
@@ -466,6 +476,32 @@ func (h *HTTPHandler) serveConfig(rw http.ResponseWriter, req *http.Request) {
+		// Either a multi-node bootstarp or a restoration of single member in multi-node.
+		m := member.NewMemberControl(h.EtcdConnectionConfig)
+
+		// check for a learner presence in a cluster.
```
```suggestion
		// check whether a learner is present in the cluster.
```

shreyas-s-rao commented on pkg/server/backuprestoreserver.go:
```diff
@@ -291,10 +292,23 @@ func (b *BackupRestoreServer) runServer(ctx context.Context, restoreOpts *brtype
+		Promote: func(ctx context.Context, logger *logrus.Entry) {
+			if restoreOpts.OriginalClusterSize > 1 {
+				m := member.NewMemberControl(b.config.EtcdConnectionConfig)
+				if err := m.PromoteMember(ctx); err == nil {
```
Should we first check whether the member is sync with the leader before attempting to promote?

ishan16696 replied:
no, `PromoteMember ` will only succeed if learner is in a sync with a leader of cluster.
So, if it gets failed then it will try again as it is in a leader-elector loop.

shreyas-s-rao commented on pkg/member/member_control.go:
```diff
@@ -265,16 +272,55 @@ func (m *memberControl) UpdateMember(ctx context.Context, cli client.ClusterClos
+
+// IsLearnerPresent checks for the learner(non-voting) member in a cluster.
+func (m *memberControl) IsLearnerPresent(ctx context.Context) (bool, error) {
+	m.logger.Infof("Removing the %v member from cluster", m.podName)
```
Remove log line

ishan16696 replied:
update the log

shreyas-s-rao commented on pkg/types/leaderelection.go:
```diff
@@ -48,7 +48,12 @@ type MemberLeaseCallbacks struct {
 
 // IsLeaderCallbackFunc is type declaration for callback function to Check LeadershipStatus.
-type IsLeaderCallbackFunc func(context.Context, *EtcdConnectionConfig, time.Duration, *logrus.Entry) (bool, error)
+type IsLeaderCallbackFunc func(context.Context, *EtcdConnectionConfig, time.Duration, *logrus.Entry) (bool, bool, error)
```
Rename this, since it now also checks for learner status, not just leadership status.

shreyas-s-rao commented on pkg/leaderelection/leaderelection.go:
```diff
@@ -143,20 +145,28 @@ func (le *LeaderElector) Run(ctx context.Context) error {
 // IsLeader checks whether the current instance of backup-restore is leader or not.
-func IsLeader(ctx context.Context, etcdConnectionConfig *brtypes.EtcdConnectionConfig, etcdConnectionTimeout time.Duration, logger *logrus.Entry) (bool, error) {
+// It also returns the boolean indicating the presence of learner(non-voting) member.
+func IsLeader(ctx context.Context, etcdConnectionConfig *brtypes.EtcdConnectionConfig, etcdConnectionTimeout time.Duration, logger *logrus.Entry) (bool, bool, error) {
```
Rename function since it no longer checks for only leadership status, but also learner status.

shreyas-s-rao commented on pkg/leaderelection/leaderelection.go:
```diff
@@ -174,15 +184,18 @@ func IsLeader(ctx context.Context, etcdConnectionConfig *brtypes.EtcdConnectionC
+		return true, false, nil
 	} else if response.Leader == NoLeaderState {
-		return false, &errors.EtcdError{
+		return false, false, &errors.EtcdError{
```
Is it not possible that a member reports `NoLeaderState` and can still be a learner?

ishan16696 replied:
`NoLeaderState ` represents that there is no quorum present in the cluster.
so, checking whether the leaner is present or not doesn't make sense as we won't be able to perform any action on learner until there will be a leader in a cluster.
so, IMO it is better to wait until there is a leader as it is in a leader-elector loop, so it will recheck and won't backup fire in any case. Hope that's make sense.

shreyas-s-rao commented on pkg/leaderelection/leaderelection_test.go:
```diff
@@ -170,5 +180,27 @@ var _ = Describe("Etcd Cluster", func() {
 			})
 		})
+
+		Context("Etcd member is leaner", func() {
```
```suggestion
		Context("Etcd member is learner", func() {
```

shreyas-s-rao reviewed (changes requested):
@ishan16696 Thanks for the well-written PR. I have added a few questions/suggestions in my review. Can you please address them? Thanks.

ishan16696 commented on pkg/initializer/initializer.go:
```diff
@@ -225,3 +230,33 @@ func (e *EtcdInitializer) removeDir(dirname string) error {
+	if err := retry.OnError(retry.DefaultBackoff, func(err error) bool {
+		return err != nil
+	}, func() error {
+		return m.RemoveMember(context.TODO())
```
I knew this review comment will come 😅 ..... I didn’t find meaning full context can you please help figuring out meaning full context to be used. 
I did some thing ...now take a look if it make sense.

shreyas-s-rao replied:
It's better now, since you're passing the context from the parent. I would still suggest to use a timeout-based context instead of a `Background` context.

ishan16696 replied:
timeout-based context for restoration doesn't seems correct to me as then next question will be  what should be the correct time ? and we have to consider the upper limit of restoration might take .... for now guessing a time is more difficult to me. WDYT?

shreyas-s-rao replied:
But the context is not for the actual restoration operation itself. It's just for the removal or adding of member, which can definitely be time-boxed. I think you can create a context(s) within the `restoreInMultiNode` method, specifically for removing and adding the members.

ishan16696 replied:
ok, but this function is calling `RemoveMember`, `AddMemberAsLearner`  and the thing is these function is using the time boxed context. Please check [here](https://github.com/gardener/etcd-backup-restore/pull/509/files#diff-e24a1a24a9ff6cdb41bc91756edd5a882dc7e91805d4305a27b82e93c5d16267R285) and [here](https://github.com/gardener/etcd-backup-restore/blob/master/pkg/member/member_control.go#L101).

ishan16696 commented on pkg/initializer/initializer.go:
```diff
@@ -225,3 +230,33 @@ func (e *EtcdInitializer) removeDir(dirname string) error {
+	if err := retry.OnError(retry.DefaultBackoff, func(err error) bool {
+		return err != nil
+	}, func() error {
+		return m.AddMemberAsLearner(context.TODO())
```
https://github.com/gardener/etcd-backup-restore/pull/509#discussion_r931778135

ishan16696 commented on pkg/initializer/initializer.go:
```diff
@@ -90,14 +88,21 @@ func (e *EtcdInitializer) Initialize(mode validator.Mode, failBelowRevision int6
+		if dataDirStatus == validator.DataDirStatusInvalidInMultiNode || (e.Validator.OriginalClusterSize > 1 && dataDirStatus == validator.DataDirectoryCorrupt) {
+			e.restoreInMultiNode()
+		} else if e.Validator.OriginalClusterSize > 1 && isPresent {
+			e.restoreInMultiNode()
```
there are no grafana is available for Multi-node, so spending time on this thing ... doesn't make sense to me.

shreyas-s-rao replied:
I think you can add it already in this PR, unless you can create a separate issue for tracking the addition of new prometheus metrics for multi-node etcd (so that we don't miss out such minor cases)

ishan16696 replied:
Isn't we already have an issue open which will take care of all the metrices required for multi-node: https://github.com/gardener/etcd-druid/issues/358

shreyas-s-rao replied:
Ok. Can we start populating that issue? Like a checklist of metrics to be added. That will ensure that we don't miss this particular one, or many others that should be part of multi-node etcd PRs.

ishan16696 replied:
make sense. done https://github.com/gardener/etcd-druid/issues/358#issuecomment-1198917054

ishan16696 commented on pkg/initializer/initializer.go:
```diff
@@ -90,14 +88,21 @@ func (e *EtcdInitializer) Initialize(mode validator.Mode, failBelowRevision int6
-			metrics.RestorationDurationSeconds.With(prometheus.Labels{metrics.LabelSucceeded: metrics.ValueSucceededTrue}).Observe(time.Since(start).Seconds())
+		if dataDirStatus == validator.DataDirStatusInvalidInMultiNode || (e.Validator.OriginalClusterSize > 1 && dataDirStatus == validator.DataDirectoryCorrupt) {
+			e.restoreInMultiNode()
+		} else if e.Validator.OriginalClusterSize > 1 && isPresent {
```
I will do it at the end as once logic will get verify from you guys.

ishan16696 replied:
done.PTAL

ishan16696 commented on pkg/initializer/initializer.go:
```diff
@@ -225,3 +230,33 @@ func (e *EtcdInitializer) removeDir(dirname string) error {
+// * Clean the data-dir of member that needs to be restored.
+// * Add a new member as a learner(non-voting member)
+// * Start a learner by providing config to etcd
+// * Promote the learner to voting member.
```
done

ishan16696 commented on pkg/miscellaneous/miscellaneous.go:
```diff
@@ -462,5 +464,39 @@ func GetInitialClusterState(ctx context.Context, logger logrus.Entry, clientSet
+	memPromoteCtx, cancel := context.WithTimeout(ctx, brtypes.DefaultEtcdConnectionTimeout)
+	defer cancel()
+
+	//Member promote call will succeed only if member is in sync with leader, and will error out otherwise
```
@aaronfern can you answer this as I guess you have better insight regarding this.

aaronfern replied:
There's no other way to ensure that a member is in sync with the leader. Right now, a member promote is the only way to check that.

ishan16696 commented on pkg/miscellaneous/miscellaneous.go:
```diff
@@ -462,5 +464,39 @@ func GetInitialClusterState(ctx context.Context, logger logrus.Entry, clientSet
 	}
 
-	return ClusterStateNew
+	return ""
```
> can't the cluster state be determined using the etcd API itself?

I don't think there is any etcd API call available which will give us cluster state as cluster state is passed to configuration of etcd. I think there is no way other then we have to determine by distinguish between each case like: `bootstrap` , `Scale-up` , `restoration`.

unmarshall replied:
```go
if errSts != nil {
		logger.Warn("error fetching etcd sts ", errSts)
		return ClusterStateNew
	}
```
Can you explain if there is an error fetching etcd sts, how is the cluster state new?

aaronfern replied:
That's more to do with default values. So the assumption is that if we can't find a sts (which we use to determine if a cluster is in scale-up) then we fall back to what is present in the etcd config, i.e. `new`

No sts present would ideally not be a case we encounter in g/g, but rather by some external party using etcd-backup-restore standalone. In which case if it's being used without an sts then some level of manual intervention will be needed to scale-up a cluster

unmarshall replied:
> No sts present would ideally not be a case we encounter in g/g, but rather by some external party using etcd-backup-restore standalone.

Not sure if this assumption is correct. There is a lot of code in `etcd-backup-restore` which assumes there is an sts. So why only in this occasion a special handling is required?

aaronfern replied:
In any case, I think we still do need some default value to fall back on in the case of any errors and defaulting to `new` seemed like a straightforward thing to do 
wdyt? do you have any other suggestion in mind?

ishan16696 commented on pkg/initializer/initializer.go:
```diff
@@ -90,14 +88,21 @@ func (e *EtcdInitializer) Initialize(mode validator.Mode, failBelowRevision int6
-			metrics.RestorationDurationSeconds.With(prometheus.Labels{metrics.LabelSucceeded: metrics.ValueSucceededTrue}).Observe(time.Since(start).Seconds())
+		if dataDirStatus == validator.DataDirStatusInvalidInMultiNode || (e.Validator.OriginalClusterSize > 1 && dataDirStatus == validator.DataDirectoryCorrupt) {
+			e.restoreInMultiNode(ctx)
+		} else if e.Validator.OriginalClusterSize > 1 && isEtcdMemberPresent {
```
yes, I will remove it

ishan16696 commented on pkg/initializer/validator/datavalidator.go:
```diff
@@ -173,6 +169,19 @@ func (d *DataValidator) sanityCheck(failBelowRevision int64) (DataDirStatus, err
 }
 
+// checkStatus checks/filter the status of sanity check.
+func (d *DataValidator) checkStatus(status DataDirStatus, err error) (DataDirStatus, error) {
```
I created this for improving the code readability. So now [here](https://github.com/gardener/etcd-backup-restore/pull/509/files#diff-52c7c484afe34d92c4af327334cfb57586bb3ace146afec66f5cd95d03bd9159R71) we don't have to worry about handling all the different scenario.

ishan16696 commented on pkg/initializer/validator/types.go:
```diff
@@ -38,9 +38,8 @@ const (
-	// TODO: To be removed when backup-restore supports restoration of single member in multi-node etcd cluster.
-	DataDirStatusUnknownInMultiNode
+	// DataDirStatusInvalidInMultiNode indicates validator failed to check the data directory status in multi-node etcd cluster.
+	DataDirStatusInvalidInMultiNode
```
sorry, didn't get you. what makes you think so .

ishan16696 commented on pkg/member/member_control.go:
```diff
@@ -265,16 +272,55 @@ func (m *memberControl) UpdateMember(ctx context.Context, cli client.ClusterClos
+		return ErrMissingMember
+	}
+
+	return m.removeMemberFromCluster(ctx, cli, foundMember.GetID())
```
so that we can able to write unit tests, as you can see `removeMemberFromCluster` takes the etcd client as an input, so the logic is to mock this etcd client and passed the mock client to this func. 
you can take look at these [unit tests](https://github.com/gardener/etcd-backup-restore/pull/509/commits/fb96427033e0276da7095e228151c16e9052f98e#diff-c999015e85f8489977cc7e1e2ef722767361b9a3524da0543443dec10f777d11R310-R369) which I added.

ishan16696 commented on pkg/member/member_control.go:
```diff
@@ -265,16 +272,55 @@ func (m *memberControl) UpdateMember(ctx context.Context, cli client.ClusterClos
+
+// IsLearnerPresent checks for the learner(non-voting) member in a cluster.
+func (m *memberControl) IsLearnerPresent(ctx context.Context) (bool, error) {
+	m.logger.Infof("checking the presence of a learner in a cluster...")
```
why I should remove this .... Isn't more logs will be beneficial in case of debugging if some issue occurs.

unmarshall replied:
If the log adds value then yes. It only logs a static string with no other information. If you wish to keep it then its fine.

ishan16696 commented on pkg/member/member_control.go:
```diff
@@ -265,16 +272,55 @@ func (m *memberControl) UpdateMember(ctx context.Context, cli client.ClusterClos
+	ctx, cancel := context.WithTimeout(ctx, brtypes.DefaultEtcdConnectionTimeout)
+	defer cancel()
+
+	return miscellaneous.CheckIfLearnerPresent(ctx, cli)
```
same reason https://github.com/gardener/etcd-backup-restore/pull/509/files#r932548975 .... I already have added unit tests for this.

ishan16696 commented on pkg/member/member_control.go:
```diff
@@ -265,16 +272,55 @@ func (m *memberControl) UpdateMember(ctx context.Context, cli client.ClusterClos
+}
+
+// removeMemberFromCluster removes member of given ID from etcd cluster.
+func (m *memberControl) removeMemberFromCluster(ctx context.Context, client client.ClusterCloser, memberID uint64) error {
```
same reason https://github.com/gardener/etcd-backup-restore/pull/509/files#r932548975

ishan16696 commented on pkg/initializer/initializer.go:
```diff
@@ -225,3 +230,31 @@ func (e *EtcdInitializer) removeDir(dirname string) error {
+// * Add a new member as a learner(non-voting member)
+func (e *EtcdInitializer) restoreInMultiNode(ctx context.Context) error {
+	m := member.NewMemberControl(e.Config.EtcdConnectionConfig)
+	if err := retry.OnError(retry.DefaultBackoff, func(err error) bool {
```
thanks for pointing this, done.

ishan16696 commented on pkg/initializer/initializer.go:
```diff
@@ -225,3 +230,31 @@ func (e *EtcdInitializer) removeDir(dirname string) error {
+		return fmt.Errorf("unable to remove the data-dir %v", err)
+	}
+
+	if err := retry.OnError(retry.DefaultBackoff, func(err error) bool {
```
done

ishan16696 commented:
Hi @unmarshall ,
I have addressed most of your review comments. I might not be able to addressed all comments as released is planned for today and I want to focus more on requirement of feature completeness and testing. Hope that's fine.

ishan16696 commented:
> Can you explain if there is an error fetching etcd sts, how is the cluster state new?

I guess this func was added during Scale-up PR. May be @aaronfern can better tell us about this as I don't have any strong opinion on this.

aaronfern commented on pkg/leaderelection/leaderelection.go:
```diff
@@ -174,15 +184,18 @@ func IsLeader(ctx context.Context, etcdConnectionConfig *brtypes.EtcdConnectionC
 	if response.Header.MemberId == response.Leader {
-		return true, nil
+		return true, false, nil
 	} else if response.Leader == NoLeaderState {
```
Ah, okay
If I understood this right, `NoLeaderState` is a state where there is no leader in the cluster. Did I get this right?

ishan16696 replied:
yes, `NoLeaderState` means quorum loss

unmarshall replied:
Is it possible that there is currently no leader but the current etcd is a learner? If yes then this method will not return correctly result, right?
Also why return an error when there is no leader?

ishan16696 replied:
> Is it possible that there is currently no leader but the current etcd is a learner?

yes, it is possible but the question is can we take any action if there is no leader present in cluster. The answer is no, we can't take any action, so it is better to wait until there is a etcd leader present in a cluster.

> Also why return an error when there is no leader?

Then we have to action like closing the snapshotter, member lease renewal ... please refer to doc:https://github.com/gardener/etcd-backup-restore/blob/master/doc/concepts/leader_election.md

ishan16696 commented on pkg/initializer/initializer.go:
```diff
@@ -90,14 +88,21 @@ func (e *EtcdInitializer) Initialize(mode validator.Mode, failBelowRevision int6
-		}
-		if restored {
-			metrics.RestorationDurationSeconds.With(prometheus.Labels{metrics.LabelSucceeded: metrics.ValueSucceededTrue}).Observe(time.Since(start).Seconds())
+		if dataDirStatus == validator.DataDirStatusInvalidInMultiNode || (e.Validator.OriginalClusterSize > 1 && dataDirStatus == validator.DataDirectoryCorrupt) {
```
But our action depends on whether it is a single node restoration in a multi-node etcd cluster or single-node restoration in a single-node etcd cluster.

unmarshall replied:
yes but the validator only validates the data directory and that is the only thing it should do. You can compose the return value from this method with any other at the consumer side. As much as possible stick to a single responsibility principle.

ishan16696 commented on pkg/member/member_control.go:
```diff
@@ -265,16 +272,55 @@ func (m *memberControl) UpdateMember(ctx context.Context, cli client.ClusterClos
+
+	foundMember := findMember(memberInfo.Members, m.podName)
+	if foundMember == nil {
+		return ErrMissingMember
```
yes, make sense ... will remove it

ishan16696 replied:
done.

ishan16696 commented on pkg/leaderelection/leaderelection.go:
```diff
@@ -143,20 +145,28 @@ func (le *LeaderElector) Run(ctx context.Context) error {
-func IsLeader(ctx context.Context, etcdConnectionConfig *brtypes.EtcdConnectionConfig, etcdConnectionTimeout time.Duration, logger *logrus.Entry) (bool, error) {
+// EtcdMemberStatus checks whether the current instance of backup-restore is leader or not.
+// It also returns the boolean indicating the presence of learner(non-voting) member.
+func EtcdMemberStatus(ctx context.Context, etcdConnectionConfig *brtypes.EtcdConnectionConfig, etcdConnectionTimeout time.Duration, logger *logrus.Entry) (bool, bool, error) {
```
Right now, I'm not sure I want to address this as right now we need a boolean to get return from `GetEtcdMemberStatus()` which also tell us about the presence of a learner (which is a follower as well).

unmarshall replied:
the third type is actually not a `Leader` but a `Learner` that is a typo. A member can in only one of the states (follower, leader, learner) therefore returning multiple values is not a clean code + design. I will leave the rest to you.

ishan16696 replied:
hmm ...let's not change the logic of leader-election here, I will address this change(if compatible) separately in this PR: https://github.com/gardener/etcd-backup-restore/pull/491. Hope that’s fine.

unmarshall replied:
okay this is fine. Can you make a note of it there?

aaronfern reviewed (changes requested):
The PR looks good from my pov
I've also tested it out and it's working fine

Just a few last minute changes :)

ishan16696 commented on pkg/initializer/initializer.go:
```diff
@@ -45,12 +46,15 @@ import (
+	ctx := context.Background()
+	var err error
+
 	//Etcd cluster scale-up case
```
I didn't get it. It's still used only by scaleup case.
I'm only using `isEtcdMemberPresent` from this func which act as a indicator for single-member restoration in multi-node.

aaronfern replied:
Ah, right
Wrong assumption from my side here

aaronfern commented on pkg/member/member_control.go:
```diff
@@ -265,16 +272,44 @@ func (m *memberControl) UpdateMember(ctx context.Context, cli client.ClusterClos
 	}
-	return err
+
+	return miscellaneous.RemoveMemberFromCluster(ctx, cli, foundMember.GetID(), &m.logger)
```
My concern here is that since we're using the same context (with a timeout) for two etcd calls, if the first one takes longer to complete for whatever reason, that will leave less time for the second call which may lead to a premature context cancellation

ishan16696 replied:
yes, I also thought that but that's why I used `DefaultEtcdConnectionTimeout =30sec` to create context.
Reasons:
1. 30secs seems enough for both API call even if one API call takes 20 secs to get complete.
2. If I used different `context.WithTimeout` for each API call then 30sec for each API call seems large to me.
3. It's more difficult to track .... let me explain:
   - you created a `context.WithTimeout` in `func A()` of 10sec. then `func A()` call `func B()`.
   - In, `func B()` you created a `context.WithTimeout` of 30 secs with parent context of `func A()`.
   - now, even if B's `context.WithTimeout` is larger/smaller ...it doesn’t matter as when A's `ctx` will get cancelled B's `ctx` will also get cancelled.
   To me it's more difficult to debug as dev might be thinking that he has given enough time to context in `func B()` but he is likely to forget about `context.WithTimeout` of parent.

unmarshall replied:
Here `ctx, cancel := context.WithTimeout(ctx, brtypes.DefaultEtcdConnectionTimeout)` the variable is shadowing the one that is passed into the function. In general this should be avoided.

If you wish to pass the same context to `RemoveMemberFromCluster` then it means that you wish to complete both `MemberList` and `RemoveMemberFromCluster` in the timeout `DefaultEtcdConnectionTimeout`. If that is the intention then the current usage is correct barring the shadowing. If that is not the intent then you should create a new context.WithTimeout and ensure that the previous one is cancelled.

aaronfern replied:
Thanks for the example @ishan16696, that makes sense. 
However in this specific case, the parent all the way up tracks to `context.Background()` and it would be beneficial to pass this parent all the way down to all the other functions and have all `context.WithTimeout` calls use this as the parent context

ishan16696 replied:
> Here ctx, cancel := context.WithTimeout(ctx, brtypes.DefaultEtcdConnectionTimeout) the variable is shadowing the one that is passed into the function. In general this should be avoided.

Ok, I will rename the ctx.

> If you wish to pass the same context to RemoveMemberFromCluster then it means that you wish to complete both MemberList and RemoveMemberFromCluster in the timeout DefaultEtcdConnectionTimeout. If that is the intention then the current usage is correct

yes, that's my intention.

ishan16696 replied:
> However in this specific case, the parent all the way up tracks to context.Background()

yes, it is.

>  it would be beneficial to pass this parent all the way down to all the other functions and have all context.WithTimeout calls use this as the parent context

why ? can you please elaborate the benefits.

aaronfern replied:
Well, passing the parent context and creating child contexts only from that will be one way of ensuring that a case like you described [here](https://github.com/gardener/etcd-backup-restore/pull/509#discussion_r936435358) won't happen


> yes, that's my intention

That's fair enough then. Since it was your intension to use the same context for both calls, and that the timeout is quite large (30s), it's good

ishan16696 replied:
I'm have now rename the ctx. PTAL

ishan16696 commented on pkg/member/member_control.go:
```diff
@@ -265,16 +272,44 @@ func (m *memberControl) UpdateMember(ctx context.Context, cli client.ClusterClos
+	}
+	defer cli.Close()
+
+	ctx, cancel := context.WithTimeout(ctx, brtypes.DefaultEtcdConnectionTimeout)
+	defer cancel()
+
```
https://github.com/gardener/etcd-backup-restore/pull/509#discussion_r936395909

ishan16696 replied:
> Will be better to track contexts

to me ... it's opposite, I feel in this way we can better track the context rather creating a context for each func call.

aaronfern replied:
But isn't creating a context within `CheckIfLearnerPresent` a better way? In that case it does not matter that the context passed to it already has a timeout
`CheckIfLearnerPresent` can then create a context with timeout that get's cancelled as soon as the func returns
Just my thought process rn

ishan16696 replied:
https://github.com/gardener/etcd-backup-restore/pull/509#discussion_r936435358

aaronfern reviewed (approved):
Looks good from my pov, thank you :) 
/lgtm

ishan16696 commented:
/ping @shreyas-s-rao

shreyas-s-rao reviewed (approved):
Thanks for making the requested changes.
/lgtm
